### ISSUE TITLE
Regression in 5.4: Excessive stack depth when comparing recursive type

### ISSUE THREAD

**Issue by cjquines** [microsoft/TypeScript#57915]
### 🔎 Search Terms

"excessive stack depth", "ts2321"

### 🕗 Version & Regression Information

- This changed in commit or PR #56004

### ⏯ Playground Link

https://www.typescriptlang.org/play?ts=5.4.3#code/C4TwDgpgBAShCGATA9gOwDYgCoFczogB4sAaKAOSggA9gJVEBnKVHAWwCMIAnMmK2vSZQAgt27wQxAHxQAvFADaAXVlyAUAEgYigOQFUAc2AALXcoF0Gzcuqj2oAfijcEKDCCUA6HzGV2HAC5YNzRMXHwiUgoyRWifLz9pAG51dVBIUXFJAFkAS1RyYjJKGithVk4eNRZ2Lm5LIRsA5zEJKSxpAODSwWsoAAYA+1bsjq6HKGDFBLgkMOw8AmKKaTIEtskZZVT1RAgAY3R4VygAMxxUA+A8tCgTeEYRYAAZBEZgFd7y5kr66QAFAFEPBgPBgnN3JhNuMSAE2AU8mx2D04QBKYIgsFQPLMGH5QpfFJAA

### 💻 Code

```ts
type ReadonlyTuple<T, N extends number, R extends Array<T> = []> =
	R['length'] extends N
    ? readonly [...R]
    : ReadonlyTuple<T, N, [T, ...R]>;

type ArrayMinN<T, N extends number> = number extends N
  ? Array<T>
  : N extends 0
    ? Array<T>
    : [...ReadonlyTuple<T, N>, ...Array<T>];

declare function hasAtLeast<T, N extends number>(
  data: ReadonlyArray<T>,
  minimum: N,
): data is ArrayMinN<T, N>;
```

### 🙁 Actual behavior

The function `hasAtLeast` does not typecheck, with an error on the return type `data is ArrayMinN<T, N>;`: "Excessive stack depth comparing types 'ArrayMinN<T, N>' and 'readonly T[]'.(2321)"

### 🙂 Expected behavior

No type error.

### Additional information about the issue

This was reported in remeda/remeda#572.

It's not clear to me whether this is working as intended, though it is a regression. It's not hard to rewrite the type to make it work, but wondering if it's a Deeper Problem™.

**Comment by llllvvuu** [comment]
Just upon a naive reading of the PR description,

> Previously, we'd obtain the constraint of IsArray<U> by applying the distributive conditional type to the constraint of U

Sounds like `ArrayMinN<T, N>` would have been constrained to `ArrayMinN<T, number>` which is just `Array<T> whereas now,

> A | B when C possibly extends X.

So it will take the union of all of your ternary branches, which will cause an infinite loop in `ReadonlyTuple`.

So it seems like this is "intended behavior" in some sense, though I don't know if it also constitutes a bug that can/should be fixed. It seems a bit hard without something inductive `Nat`. Note that even if TypeScript were smarter, `N = 3.5` would still infinite loop if I'm reading this correctly.

**Comment by RyanCavanaugh** [comment]
Unfortunately, whether or not you hit this limiter is very dependent on the exact order that things happen internally, so implementation details can cause this to happen or not happen pretty easily.
